md5.c

```diff
@@ -1,277 +1,358 @@
+ */
+static void byteReverse(unsigned char *buf, unsigned longs)
+{
+    uint32 t;
+    do {
+	t = (uint32) ((unsigned) buf[3] << 8 | buf[2]) << 16 |
+	    ((unsigned) buf[1] << 8 | buf[0]);
+	*(uint32 *) buf = t;
+	buf += 4;
+    } while (--longs);
+}
+
+/*
+ * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
+ * initialization constants.
+ */
+void rpmMD5Init(struct MD5Context *ctx)
+{
+    ctx->buf[0] = 0x67452301;
+    ctx->buf[1] = 0xefcdab89;
+    ctx->buf[2] = 0x98badcfe;
+    ctx->buf[3] = 0x10325476;
+
+    ctx->bits[0] = 0;
+    ctx->bits[1] = 0;
+
+    if (IS_BIG_ENDIAN()) {
+	ctx->doByteReverse = 1;
+    } else {
+	ctx->doByteReverse = 0;
+    }
+}
+
+/*
+ * Update context to reflect the concatenation of another buffer full
+ * of bytes.
+ */
+void rpmMD5Update(struct MD5Context *ctx, unsigned char const *buf, unsigned len)
+{
+    uint32 t;
+
+    /* Update bitcount */
+
+    t = ctx->bits[0];
+    if ((ctx->bits[0] = t + ((uint32) len << 3)) < t)
+	ctx->bits[1]++;		/* Carry from low to high */
+    ctx->bits[1] += len >> 29;
+
+    t = (t >> 3) & 0x3f;	/* Bytes already in shsInfo->data */
+
+    /* Handle any leading odd-sized chunks */
+
+    if (t) {
+	unsigned char *p = (unsigned char *) ctx->in + t;
+
+	t = 64 - t;
+	if (len < t) {
+	    memcpy(p, buf, len);
+	    return;
+	}
+	memcpy(p, buf, t);
+	if (ctx->doByteReverse)
+	    byteReverse(ctx->in, 16);
+	rpmMD5Transform(ctx->buf, (uint32 *) ctx->in);
+	buf += t;
+	len -= t;
+    }
+    /* Process data in 64-byte chunks */
+
+    while (len >= 64) {
+	memcpy(ctx->in, buf, 64);
+	if (ctx->doByteReverse)
+	    byteReverse(ctx->in, 16);
+	rpmMD5Transform(ctx->buf, (uint32 *) ctx->in);
+	buf += 64;
+	len -= 64;
+    }
+
+    /* Handle any remaining bytes of data. */
+
+    memcpy(ctx->in, buf, len);
+}
+
+/*
+ * Decompress (BZIP) the data, then update context as per usual.
+ */
+void rpmMD5DecompressUpdate(struct MD5Context *ctx, unsigned char const *buf, unsigned len)
+{
+    int ret, used;
+
+    bz_stream addbz2strm;
+    addbz2strm.bzalloc = NULL;
+    addbz2strm.bzfree = NULL;
+    addbz2strm.opaque = NULL;
+    addbz2strm.next_in = buf;
+    addbz2strm.avail_in = len;
+
+    if (BZ2_bzDecompressInit(&addbz2strm, 0, 0) != BZ_OK)
+    {
+      fprintf(stderr, "MD5Update: BZ2_bzDecompressInit error\n");
+      exit(1);
+    }
+
+    do {
+      unsigned char *out = (char*)xmalloc(len);
+      addbz2strm.next_out = out;
+      addbz2strm.avail_out = sizeof(out);
+
+      used = addbz2strm.avail_out;
+      ret = BZ2_bzDecompress(&addbz2strm);
+      used -= addbz2strm.avail_out;
+
+      rpmMD5Update(ctx, out, used);
+      xfree(out);
+
+      if (ret == BZ_STREAM_END) {
+        break;
+      }
+    } while (addbz2strm.avail_out == 0);
+
+    BZ2_bzDecompressEnd(&addbz2strm);
+}
+
+/*
+ * Inflate (GZIP) data in buf, then update context as per usual.
+ */
+void rpmMD5InflateUpdate(struct MD5Context *ctx, unsigned char const *buf, unsigned len)

```

Comment:
New method added for inflating GZIP data before calling the usual rpmMD5Update()